src/gui/editorwidgets/core/qgseditorwidgetwrapper.cpp

```diff
@@ -92,3 +94,70 @@ void QgsEditorWidgetWrapper::valueChanged()
 {
   emit valueChanged( value() );
 }
+
+void QgsEditorWidgetWrapper::updateConstraintWidgetStatus()
+{
+  if ( mValidConstraint )
+    widget()->setStyleSheet( "" );
+  else
+    widget()->setStyleSheet( "background-color: #dd7777;" );
+}
+
+void QgsEditorWidgetWrapper::updateConstraint( const QgsFeature &ft )
+{
+  bool toEmit( false );
+  QString errStr( "predicate is True" );
+  QString expression = layer()->editFormConfig()->constraint( mFieldIdx );
+  QVariant value = ft.attribute( mFieldIdx );
+
+  if ( ! expression.isEmpty() )
+  {
+    QgsExpressionContext context =
```

Comment:
Can you also add the layer scope here? ie

` context << QgsExpressionContextUtils::layerScope( layer() );`

Then layer related variables will also be usable here.